src/client/utils/teleporter/teleporter.tsx

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017-2019 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import { Unary } from "../../../common/utils/functional/functional";
+
+interface Context {
+  node?: Element;
+  notify?: Unary<Element, void>;
+}
+
+interface SourceState {
+  node: Element;
+}
+
+interface Teleporter {
+  Source: typeof React.Component;
```

Comment:
Looks bad, but I didn't manage to get anything better.

Reply:
OTOH, consumer don't care what state component holds. And we expect no props at all, sooo...